parser/src/packet2.rs

```diff
@@ -319,30 +306,19 @@ impl<'argtype> Parser<'argtype> {
         if zero != 0 {
             panic!("What does this field mean?");
         }
-        let (i, x) = le_f32(i)?;
-        let (i, y) = le_f32(i)?;
-        let (i, z) = le_f32(i)?;
-        let (i, rot_x) = be_u32(i)?;
-        let (i, rot_y) = be_u32(i)?;
-        let (i, rot_z) = be_u32(i)?;
-        let (i, a) = le_f32(i)?;
-        let (i, b) = le_f32(i)?;
-        let (i, c) = le_f32(i)?;
-        let (i, extra) = be_u8(i)?;
+        let (i, position) = Vector3::parse(i).unwrap();
```

Comment:
Okay, my grasp on nom error handling is a little flimsy, but I think I've at least figured out the problem. (note, that my own flimsy understanding probably means that the `Error` and `ErrorKind` types are setup in a way that makes this hard, and there's a way to make those structs such that this problem doesn't exist)

So to summarize the issue, at a high level this function returns an `nom::IResult<_, _, crate::error::Error>`, which means it's a `Result<_, nom::Err<crate::error::Error>>`. For most builtin nom parsers, like `le_u32`, the return type is generic over error - so as long as `E` in `Result<_, nom::Err<E>>` implements `ParserError` then `le_u32` will directly return that type and it's fine.

However, `nom-derive` I believe hardcodes its error return type to be `nom::Err<nom::error::Error>` (because it hardcodes the return type to be `nom::IResult`). So, `Vector3::parse(i)` is returning a `Result<_, nom::Err<nom::error::Error>>` which the `?` operator will try to convert into `Result<_, nom::Err<crate::error::Error>>`, which requires that `nom::Err<crate::error::Error>` implement `From<nom::Err<nom::error::Error>>`. Which is the error you get if you try that.

But, the problem is that you can't implement that from trait, because both the types are foreign.

So I think there's three possible fixes:
* Change `nom-derive` to generate code that is generic over type, or
* Change `nom` such that `nom::Err<E1>` implements `From<nom::Err<E2>>` iff `E1` implements `From<E2>`, or
* Hand-implement `Parse` for `Vector3` and `Rotation3`.

The first two, obviously, don't help us now since we'd have to change upstream code. The last one is kinda tedious though.

I'll leave it to you which approach you want to take. I don't expect `unwrap`ing to be a huge problem in practice, I just ran this change against my replay dataset (~2K replays since version 0.9.10) and it worked fine.

Reply:
3 is by far the smartest, will update the pr when i get time